src/ast/nodes/shared/FunctionNode.ts

```diff
@@ -83,12 +74,14 @@ export default class FunctionNode extends NodeBase {
 
 	initialise() {
 		this.included = false;
+		this.isPrototypeReassigned = false;
 		if (this.id !== null) {
 			this.id.declare('function', this);
 		}
 		for (const param of this.params) {
 			param.declare('parameter', null);
 		}
+		this.body.addImplicitReturnExpressionToScope();
```

Comment:
It seems here's an example case that is invalided by this change:

```js
function x () {}
function y () {}
function f () {
  if (Math.random() > 0.5)
    return y;
  return x;
}
f()();
```

that would otherwise be tree-shaken.

It doesn't seem a huge loss to me though as it is really hypothetical.

I wonder though if we could not treat the whole function itself as having a "single type" (set from the first return), and only opt-out only if we find that a return value doesn't follow that type rule? That might stop the deep loops while still retaining the basic functionality here? Just a thought.

Reply:
I guess you wouldn't know that `x` and `y` are the same without the full deep analysis though... so yeah. But perhaps having a type like `f: PURE_FUNCTION` or `f: STRING_LITERAL` would be useful?

Reply:
It might be but it would also mean creating a whole new analysis step and another layer of types. I would rather wait for us to have some basic SSA going and use this as a base for much better analysis. Otherwise I would not take action unless there are some situations in important libraries for which this change would provide important gains.

As related to your first comment, rollup also does not yet detect unreachable code after return, break or throw statements. So this would be treated as "unknown" as well:
```javascript
function x() {
  return false;
  const unimportant = false;
}

if (x()) console.log('not removed');
```

Reply:
Sure, as I say these certainly seem suitable compromises to me.